### ISSUE TITLE
Performance Efficiency Lab "Calculating Differences In Clock Sources" Needs Updating

### ISSUE THREAD

**Issue by jolson629** [awslabs/aws-well-architected-labs#128]
In the 100 level lab for Performance Efficiency, ["Calculating Differences In Clock Sources"](https://www.wellarchitectedlabs.com/performance-efficiency/100_labs/100_clock_source_performance/), the lab is predicated on using the AWS Systems Manager to execute scripts on two different machine type to collect performance counters.

However, I think AWS Systems Manager (SM) has changed siginificantly since this lab was created, and the procedure outlined in step 2: "test performance" is not valid.

In step 2, you are supposed to use the 'run command' functionality of SM to execute a script on both machines setup by the lab's CloudFormation script. You first select the script, then target the machines (manually) to execute the script on:

<img width="1416" alt="Screen Shot 2020-08-27 at 11 00 47 AM" src="https://user-images.githubusercontent.com/67236918/91466349-9bfa7200-e854-11ea-806e-6657eb973ffd.png">

However, there is no mention about first registering the instance with SM in the lab documentation, which needs to be done prior to selecting them for a 'run command'. You can see the user interface has changed between the lab documentation and the current UI:

<img width="1592" alt="Screen Shot 2020-08-27 at 11 02 23 AM" src="https://user-images.githubusercontent.com/67236918/91466988-89346d00-e855-11ea-8a3f-1ad932231fea.png">

Following the procedure to register EC2 instance with SM, as documented in the 100 lab "Inventory and Patch Management" under Operational Efficiency proved unsuccessful as well...Do the AMIs the EC2 instances this lab uses support the SSM Agent SM requires? I could never get the EC2 instances to register with SM, even assigning the EC2 instances the correct IAM role for SM.

I love the idea of this lab, and would love to see it to completion!

**Comment by ericpullen** [comment]
This looks like an issue with the VPC you are deploying this into. I didn't make it clear in the lab (I will add it), but the VPC needs to be setup to allow SSM to work. Did you deploy this into the default VPC? If so, it should have worked if there were not changes to the base VPC NACL's and such. 

As a test, can you try deploying the same lab to a different region (select the region from the dropdown in the upper left) and select the default VPC.  If that works, then we know it was an issue with the VPC configuration.

**Comment by ericpullen** [comment]
I am going to add the following warning to the deploy step of the lab:

If you have modified the default VPC or are using a VPC you have created, ensure that the subnet you are deploying the EC2 instances into can communicate with AWS Systems Manager. One method for this is to [Create a VPC endpoint for SSM](https://docs.aws.amazon.com/systems-manager/latest/userguide/setup-create-vpc.html)

**Comment by jolson629** [comment]
Getting closer...adding the three SSM PrivateLink endpoints to the VPC managing the subnet I used as a param on the CloudFormation script did the trick, although I am a little hazy on why a default VPC would already have these endpoints enabled. I did learn from the lab CloudFormation script how to get the EC2 instances to automatically register with SSM...I didn't know how to do that before.

Onto the next error...After executing the 'Run Command' step (step 8) in the lab, the run command exited with status 'Failed' on both machines:

<img width="1873" alt="Screen Shot 2020-09-04 at 8 21 24 AM" src="https://user-images.githubusercontent.com/67236918/92244479-98846d80-ee88-11ea-8978-6b17e9a08cbb.png">

Drilling down into both machines, they have the same error:

<img width="1874" alt="Screen Shot 2020-09-04 at 8 20 45 AM" src="https://user-images.githubusercontent.com/67236918/92244563-b4880f00-ee88-11ea-840b-6a6f567107d3.png">

**Comment by jolson629** [comment]
Cloud init logs uploaded from one of the two EC2 machine. 

It looks like this lab is expecting the subnet that the EC2 machines are being launched into has internet access - hence the problems with the SSM endpoints above, and the errors in the logs below.

You might want to make a note of that in the lab setup guide. If you are running this lab in a corporate environment, your test area for stuff like this may not have internet access by default to yum install needed package...


[cfn-init.log](https://github.com/awslabs/aws-well-architected-labs/files/5176793/cfn-init.log)
[cloud-init.log](https://github.com/awslabs/aws-well-architected-labs/files/5176794/cloud-init.log)
[cloud-init-output.log](https://github.com/awslabs/aws-well-architected-labs/files/5176795/cloud-init-output.log)

**Comment by ericpullen** [comment]
I have altered the language to now read: 
If you have modified the default VPC or are using a VPC you have created, ensure that the subnet you are deploying the EC2 instances into can communicate with the internet and with AWS Systems Manager. One method for this is to Create a VPC endpoint for SSM.  Hopefully that will make it clear for future usage. 

There is no PrivateLink, but it is possible to access the yum repositories within a VPC without access to the internet using a VPC endpoint for S3. Your VPC endpoint policy must allow traffic from your VPC to the S3 buckets that host the Amazon Linux AMI repositories.
Below is an example VPC endpoint policy that accomplishes this:
  DemoVpcS3Endpoint:
    Type: AWS::EC2::VPCEndpoint
    Properties:
      PolicyDocument:
        Version: 2012-10-17
        Statement:
          - Effect: Allow
            Principal: '*'
            Action:
              - 's3:GetObject'
            Resource:
              - !Sub 'arn:aws:s3:::amazonlinux.${AWS::Region}.amazonaws.com/*'


Try that and see if it works.

**Comment by jolson629** [comment]
Ok, I have this lab working 100% end-to-end without using the public internet. I'm going to send a pull request over that includes an update CloudFormation .yaml file for your review. This script creates all the needed lab artifacts from the ground up, including the VPC, the gateway VPC endpoint (S3),and the interface VPC endpoints (SSM, SSM Messages, EC2 Messages, CloudFormation) with the needed policies and security groups to run the lab.

I got **a lot more** than I bargained for out of this lab! Hopefully I can help someone else trying to execute in an internet-less environment, or someone that wants to learn how the VPC endpoints work in lieu of. 

Thanks for the help, Eric...I learned a lot here.

Consider my issue here closed.
